File: scala-java/src/main/java/at/twinformatics/scalademo/java/Iterator.java
package at.twinformatics.scalademo.java;

/**
 * Defining a simple interface in Java, including use of default method.
 */
public interface Iterator<T> {
    boolean hasNext();
    T next();
    default void remove() {
        throw new UnsupportedOperationException();
    }
}

File: scala-java/src/main/scala/at/twinformatics/scalademo/scala/Iterator.scala
package at.twinformatics.scalademo.scala

/**
  * As simple example on how to define traits. Traits doesn't require an
  * "abstract" keyword. Just leave out the implementation. The same applies to
  * "default" methods, just give a method implementation.
  */
trait Iterator[T] {
    def hasNext: Boolean
    def next: T
    def remove(): Unit = {
        throw new UnsupportedOperationException()
    }
}
